include/eve/detail/function/bit_cast.hpp

```diff
@@ -21,6 +21,15 @@ namespace eve
     {
       return EVE_DISPATCH_CALL(a,tgt);
     }
+    // In riscv we may have different underlay type sizes
+#if defined(EVE_INCLUDE_RISCV_HEADER)
+    template<typename T, typename Target>
+    requires(sizeof(T) != sizeof(Target))
+    EVE_FORCEINLINE Target operator()(T const& a, as<Target> const& tgt) const noexcept
+    {
+      return EVE_DISPATCH_CALL(a, tgt);
+    }
+#endif
```

Comment:
This doesn't make any sense to me. bitcast - means specifically to copy to the same size. That what it does.

What is the problem you are solving.

UPD: I think I see - you want to bitcast between a smaller and larger cardinals. But, I guess, they come out as different size. 

Why do you need to bitcast between them? If the spec says we have all those small registers, why not keep them?

Reply:
> Why do you need to bitcast between them?

There are two main reasons. 

First and the main one is when I need to bitcast between, for example, `wide<char, 4>` and `wide<int, 1>`. They have the same `sizeof(T) * N`, however for RISC-V underlying type sizes will be different:
 * `wide<char, 4>` could be represented as a part of real vector register in RISC-V vector intrinsics (for vlen=128 this will be 1/4 of real vector register).
 * For `wide<int, 1>` we have an ability to represent it with vlen=128 as 1/4 of vector register. However,  RISC-V vector intrinsics have some restrictions: they are unable to set LMUL to MF4 (1/4) if elements are of type int. It seems that this restriction comes from the fact that vector intrinsics are expected to work on any expected (by intrinsics) vector unit configuration. For example, if vlen=64 int will not fit into 1/4 part of vector. So, in intrinsics interface we need to use the type that corresponds to at least 1/2 of real vector register.
 
 That is why I need to "increase" LMUL, so underlying sizes will be different.

The second one is that sometimes I need an ability to cast between `wide` with different cardinal sizes, but the same types (for example `test\test.h`). In EVE it seems you usually just get a `.storage()`, and construct from it a new `wide` with increased cardinal. For RISC-V this is unavailable (as underlying types may be different), so for now I've decided to make bit_cast do this thing too. If there is better approach for this, I can try to refactor this.

Reply:
I see - what you are saying is that bit_casting from wide<char, 4> to wide<int, 1> should work, regardless of the type.

I'd agree with that - but this requires is very wrong. It'd have to say "sizeof(T) * N1::value == sizeof(U) * N2::value".

I'll have a look this weekend to make that change for you in master.

Reply:
Thanks.

Also, how should `bit_cast` be used on `logical`s?

I've used mask types (i.e. a pack of bits) as `logical`s for RISC-V.  Is `bit_cast` from `wide<char, 1>` to `logical<wide<int, 8>>` allowed (since sizes are equal)?

I ask since for some functions none of the overloads suited RISC-V (so I have written RISC-V specific ones).
E.G.: https://github.com/jfalcou/eve/blob/c1f74e3b5fe19a539bff8c20d05ba87192224e04/include/eve/detail/function/simd/common/bitmask.hpp#L50
Casts between `eve::logical<eve::wide<T, eve::fixed<N>>>` and `eve::wide<T, eve::fixed<N>>>`, but for RISC-V these types differ in length.